db/sqlglue.c

```diff
@@ -8473,25 +8466,41 @@ int sqlite3BtreeInsert(
     const BtreePayload *pPayload, /* The key and data of the new record */
     int bias, int seekResult, int flags)
 {
-    const void *pKey = pPayload->pKey;
-    sqlite3_int64 nKey = pPayload->nKey;
-    const void *pData = pPayload->pData;
-    int nData = pPayload->nData;
+    const void *pKey;
+    sqlite3_int64 nKey;
+    const void *pData;
+    int nData;
+    blob_buffer_t *pblobs;
 
     int rc = UNIMPLEMENTED;
     int bdberr;
     blob_buffer_t blobs[MAXBLOBS];
     struct sql_thread *thd = pCur->thd;
     struct sqlclntstate *clnt = pCur->clnt;
 
+    /* If pPayload is NULL, ondisk_buf and ondisk_blobs contain comdb2 row data
+       and are ready to be inserted as is. */
+    if (pPayload == NULL) {
+        pblobs = pCur->ondisk_blobs;
+        pKey = NULL;
+        nKey = 0;
+        pData = NULL;
+        nData = 0;
+    } else {
+        pKey = pPayload->pKey;
+        nKey = pPayload->nKey;
+        pData = pPayload->pData;
+        nData = pPayload->nData;
+        bzero(blobs, sizeof(blobs));
+        pblobs = blobs;
```

Comment:
Don't know how often this path will be chosen, but we could also do pblobs = alloca(); here instead of having blobs[] on the stack always.

Reply:
With my changes, this code path will be hit much less. I've made the optimization per your suggestion. Thank you!